pkg/cache/controller_test.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestClusterIndexFunc(t *testing.T) {
+	tests := []struct {
+		obj     interface{}
+		desired string
+	}{
+		{obj: &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{ClusterName: "test"}}, desired: "test//"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desired, func(t *testing.T) {
+			result, err := ClusterIndexFunc(tt.obj)
+			if err != nil {
+				t.Error(err)
+			}
+			if result[0] != tt.desired {
+				t.Errorf("got %v, want %v", result[0], tt.desired)
+			}
+
+			key, err := ClusterAwareKeyFunc(tt.obj)
+			if err != nil {
+				t.Error(err)
+			}
+
+			if result[0] != key {
+				t.Errorf("Index and Keyfunc have diverged, got %v, want %v", result[0], key)
+			}
+		})
+	}
+}
+
+func TestClusterAndNamespaceIndexFunc(t *testing.T) {
+	tests := []struct {
+		obj     interface{}
+		desired string
+	}{
+		{obj: &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{ClusterName: "test"}}, desired: "test//"},
+		{obj: &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{ClusterName: "test", Namespace: "testing"}}, desired: "test/testing/"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desired, func(t *testing.T) {
+			result, err := ClusterAndNamespaceIndexFunc(tt.obj)
+			if err != nil {
+				t.Error(err)
+			}
+			if result[0] != tt.desired {
+				t.Errorf("got %v, want %v", result[0], tt.desired)
+			}
+
+			key, err := ClusterAwareKeyFunc(tt.obj)
+			if err != nil {
+				t.Error(err)
+			}
+
+			if result[0] != key {
+				t.Errorf("Index and Keyfunc have diverged, got %v, want %v", result[0], key)
+			}
+		})
+	}
+}
+
+func TestClusterAwareKeyFunc(t *testing.T) {
+	tests := []struct {
+		obj     interface{}
+		desired string
+	}{
+		{obj: &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{ClusterName: "cluster"}}, desired: "cluster//"},
```

Comment:
when writing up these tests I realized I wasn't sure if this was the formatting we want, do we want leave the trailing `/` when namespace/name aren't provided?

Reply:
We won't ever have a key that is just cluster name. We could have cluster+name or cluster+namespace+name. So it's possible that you'd see `cluster//name` (double slashes / no namespace) for a cluster-scoped resource. We can always adjust the key format later if we need to.